In [1]:
import pandas as pd
houses = pd.read_csv("data/kc_house_data.csv")

## 2. Min, Max, and Sum

In [2]:
houses.min()
# operate on every colomn

id                       1000102
date             20140502T000000
price                    75000.0
bedrooms                       0
bathrooms                    0.0
sqft_living                  290
sqft_lot                     520
floors                       1.0
waterfront                     0
view                           0
condition                      1
grade                          1
sqft_above                   290
sqft_basement                  0
yr_built                    1900
yr_renovated                   0
zipcode                    98001
lat                      47.1559
long                    -122.519
sqft_living15                399
sqft_lot15                   651
dtype: object

In [3]:
houses.max()

id                    9900000190
date             20150527T000000
price                  7700000.0
bedrooms                      33
bathrooms                    8.0
sqft_living                13540
sqft_lot                 1651359
floors                       3.5
waterfront                     1
view                           4
condition                      5
grade                         13
sqft_above                  9410
sqft_basement               4820
yr_built                    2015
yr_renovated                2015
zipcode                    98199
lat                      47.7776
long                    -121.315
sqft_living15               6210
sqft_lot15                871200
dtype: object

In [4]:
type(houses.max())

pandas.core.series.Series

The data type of max and min methods is series.

In [5]:
houses.sum()

id                                                  98994056770455
date             20141013T00000020141209T00000020150225T0000002...
price                                                11672925008.0
bedrooms                                                     72854
bathrooms                                                 45706.25
sqft_living                                               44952873
sqft_lot                                                 326506890
floors                                                     32296.5
waterfront                                                     163
view                                                          5064
condition                                                    73688
grade                                                       165488
sqft_above                                                38652488
sqft_basement                                              6300385
yr_built                                                  4259

In [6]:
# numeric_only: Include only float, int, boolean columns. 
houses.sum(numeric_only = True)

id               9.899406e+13
price            1.167293e+10
bedrooms         7.285400e+04
bathrooms        4.570625e+04
sqft_living      4.495287e+07
sqft_lot         3.265069e+08
floors           3.229650e+04
waterfront       1.630000e+02
view             5.064000e+03
condition        7.368800e+04
grade            1.654880e+05
sqft_above       3.865249e+07
sqft_basement    6.300385e+06
yr_built         4.259933e+07
yr_renovated     1.824186e+06
zipcode          2.119759e+09
lat              1.027915e+06
long            -2.641409e+06
sqft_living15    4.293536e+07
sqft_lot15       2.759646e+08
dtype: float64

In [7]:
titanic = pd.read_csv("data/titanic.csv")

In [8]:
titanic.head(3)

pclass  survived                            name     sex     age  sibsp  \
0       1         1   Allen, Miss. Elisabeth Walton  female      29      0   
1       1         1  Allison, Master. Hudson Trevor    male  0.9167      1   
2       1         0    Allison, Miss. Helen Loraine  female       2      1   

   parch  ticket      fare    cabin embarked boat body  \
0      0   24160  211.3375       B5        S    2    ?   
1      2  113781    151.55  C22 C26        S   11    ?   
2      2  113781    151.55  C22 C26        S    ?    ?   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON

In [9]:
titanic.sum()

pclass                                                    3004
survived                                                   500
name         Allen, Miss. Elisabeth WaltonAllison, Master. ...
sex          femalemalefemalemalefemalemalefemalemalefemale...
age          290.91672302548633953714718242680?245032363747...
sibsp                                                      653
parch                                                      504
ticket       2416011378111378111378111378119952135021120501...
fare         211.3375151.55151.55151.55151.5526.5577.958305...
cabin        B5C22 C26C22 C26C22 C26C22 C26E12D7A36C101?C62...
embarked     SSSSSSSSSCCCCSSSCCCCSSCCSCCCSSSCSSSCSSSCCCSCCS...
boat         211???310?D??496B??68A55548?778D?788?469???6D8...
body         ???135?????22124??????????????148?????????????...
home.dest    St Louis, MOMontreal, PQ / Chesterville, ONMon...
dtype: object

In [10]:
titanic.sum(numeric_only = True)

pclass      3004
survived     500
sibsp        653
parch        504
dtype: int64

exclude some numbers

In [15]:
titanic.head(1306).tail(5)
# 1306的前五个

pclass  survived                   name     sex   age  sibsp  parch  \
1301       3         0   Youseff, Mr. Gerious    male  45.5      0      0   
1302       3         0      Yousif, Mr. Wazli    male     ?      0      0   
1303       3         0  Yousseff, Mr. Gerious    male     ?      0      0   
1304       3         0   Zabour, Miss. Hileni  female  14.5      1      0   
1305       3         0  Zabour, Miss. Thamine  female     ?      1      0   

     ticket     fare cabin embarked boat body home.dest  
1301   2628    7.225     ?        C    ?  312         ?  
1302   2647    7.225     ?        C    ?    ?         ?  
1303   2627  14.4583     ?        C    ?    ?         ?  
1304   2665  14.4542     ?        C    ?  328         ?  
1305   2665  14.4542     ?        C    ?    ?         ?

In [16]:
titanic.tail(5).head(3)
# 最后5行中的前3行

pclass  survived                       name     sex   age  sibsp  parch  \
1304       3         0       Zabour, Miss. Hileni  female  14.5      1      0   
1305       3         0      Zabour, Miss. Thamine  female     ?      1      0   
1306       3         0  Zakarian, Mr. Mapriededer    male  26.5      0      0   

     ticket     fare cabin embarked boat body home.dest  
1304   2665  14.4542     ?        C    ?  328         ?  
1305   2665  14.4542     ?        C    ?    ?         ?  
1306   2656    7.225     ?        C    ?  304         ?

In [17]:
titanic.tail(5).head(3).sum(numeric_only = True)

pclass      9
survived    0
sibsp       2
parch       0
dtype: int64

In [18]:
titanic.count()

pclass       1309
survived     1309
name         1309
sex          1309
age          1309
sibsp        1309
parch        1309
ticket       1309
fare         1309
cabin        1309
embarked     1309
boat         1309
body         1309
home.dest    1309
dtype: int64

Count non-NA cells for each column or row.

## 2. Mean, Median, and Mode

In [24]:
houses.mean(numeric_only=True)  # 平均值

id               4.580302e+09
price            5.400881e+05
bedrooms         3.370842e+00
bathrooms        2.114757e+00
sqft_living      2.079900e+03
sqft_lot         1.510697e+04
floors           1.494309e+00
waterfront       7.541757e-03
view             2.343034e-01
condition        3.409430e+00
grade            7.656873e+00
sqft_above       1.788391e+03
sqft_basement    2.915090e+02
yr_built         1.971005e+03
yr_renovated     8.440226e+01
zipcode          9.807794e+04
lat              4.756005e+01
long            -1.222139e+02
sqft_living15    1.986552e+03
sqft_lot15       1.276846e+04
dtype: float64

DataFrame.mean(axis=0, skipna=True, numeric_only=False, **kwargs)

In [26]:
houses.median(numeric_only=True)  # 中位数

id               3.904930e+09
price            4.500000e+05
bedrooms         3.000000e+00
bathrooms        2.250000e+00
sqft_living      1.910000e+03
sqft_lot         7.618000e+03
floors           1.500000e+00
waterfront       0.000000e+00
view             0.000000e+00
condition        3.000000e+00
grade            7.000000e+00
sqft_above       1.560000e+03
sqft_basement    0.000000e+00
yr_built         1.975000e+03
yr_renovated     0.000000e+00
zipcode          9.806500e+04
lat              4.757180e+01
long            -1.222300e+02
sqft_living15    1.840000e+03
sqft_lot15       7.620000e+03
dtype: float64

In [28]:
titanic.mode()  # most frequent value

pclass  survived                  name   sex  age  sibsp  parch    ticket  \
0     3.0       0.0  Connolly, Miss. Kate  male    ?    0.0    0.0  CA. 2343   
1     NaN       NaN      Kelly, Mr. James   NaN  NaN    NaN    NaN       NaN   

   fare cabin embarked boat body home.dest  
0  8.05     ?        S    ?    ?         ?  
1   NaN   NaN      NaN  NaN  NaN       NaN

## 3. Describe with Numeric Value

`DataFrame.describe(percentiles=None, include=None, exclude=None)[source]`
Generate descriptive statistics.
Descriptive statistics include those that summarize the central tendency, dispersion and shape of a dataset’s distribution, excluding NaN values.

In [29]:
titanic.describe()

pclass     survived        sibsp        parch
count  1309.000000  1309.000000  1309.000000  1309.000000
mean      2.294882     0.381971     0.498854     0.385027
std       0.837836     0.486055     1.041658     0.865560
min       1.000000     0.000000     0.000000     0.000000
25%       2.000000     0.000000     0.000000     0.000000
50%       3.000000     0.000000     0.000000     0.000000
75%       3.000000     1.000000     1.000000     0.000000
max       3.000000     1.000000     8.000000     9.000000

std: standard deviation

In [31]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   pclass     1309 non-null   int64 
 1   survived   1309 non-null   int64 
 2   name       1309 non-null   object
 3   sex        1309 non-null   object
 4   age        1309 non-null   object
 5   sibsp      1309 non-null   int64 
 6   parch      1309 non-null   int64 
 7   ticket     1309 non-null   object
 8   fare       1309 non-null   object
 9   cabin      1309 non-null   object
 10  embarked   1309 non-null   object
 11  boat       1309 non-null   object
 12  body       1309 non-null   object
 13  home.dest  1309 non-null   object
dtypes: int64(4), object(10)
memory usage: 143.3+ KB


In [32]:
titanic.describe(include=["object"])

name   sex   age    ticket  fare cabin embarked  boat  \
count                   1309  1309  1309      1309  1309  1309     1309  1309   
unique                  1307     2    99       929   282   187        4    28   
top     Connolly, Miss. Kate  male     ?  CA. 2343  8.05     ?        S     ?   
freq                       2   843   263        11    60  1014      914   823   

        body home.dest  
count   1309      1309  
unique   122       370  
top        ?         ?  
freq    1188       564

1. there are 1307 unique names, but 3 people have the same name.
2. top: most frequently
3. freq: the frequency of values in top occurs

In [ ]:
titanic.describe(include = ["O"])  # shorter for object. Uppercase

In [33]:
titanic.describe(include = ["O", "int64"])

pclass     survived                  name   sex   age  \
count   1309.000000  1309.000000                  1309  1309  1309   
unique          NaN          NaN                  1307     2    99   
top             NaN          NaN  Connolly, Miss. Kate  male     ?   
freq            NaN          NaN                     2   843   263   
mean       2.294882     0.381971                   NaN   NaN   NaN   
std        0.837836     0.486055                   NaN   NaN   NaN   
min        1.000000     0.000000                   NaN   NaN   NaN   
25%        2.000000     0.000000                   NaN   NaN   NaN   
50%        3.000000     0.000000                   NaN   NaN   NaN   
75%        3.000000     1.000000                   NaN   NaN   NaN   
max        3.000000     1.000000                   NaN   NaN   NaN   

              sibsp        parch    ticket  fare cabin embarked  boat  body  \
count   1309.000000  1309.000000      1309  1309  1309     1309  1309  1309   
unique          NaN          NaN       929   282   187        4    28   122   
top             NaN          NaN  CA. 2343  8.05     ?        S     ?     ?   
freq            NaN          NaN        11    60  1014      914   823  1188   
mean       0.498854     0.385027       NaN   NaN   NaN      NaN   NaN   NaN   
std        1.041658     0.865560       NaN   NaN   NaN      NaN   NaN   NaN   
min        0.000000     0.000000       NaN   NaN   NaN      NaN   NaN   NaN   
25%        0.000000     0.000000       NaN   NaN   NaN      NaN   NaN   NaN   
50%        0.000000     0.000000       NaN   NaN   NaN      NaN   NaN   NaN   
75%        1.000000     0.000000       NaN   NaN   NaN      NaN   NaN   NaN   
max        8.000000     9.000000       NaN   NaN   NaN      NaN   NaN   NaN   

       home.dest  
count       1309  
unique       370  
top            ?  
freq         564  
mean         NaN  
std          NaN  
min          NaN  
25%          NaN  
50%          NaN  
75%          NaN  
max          NaN